cpp/src/arrow/util/future_test.cc

```diff
@@ -106,12 +107,18 @@ template <typename T>
 class SimpleExecutor {
  public:
   explicit SimpleExecutor(int nfutures)
-      : pool_(ThreadPool::Make(/*threads=*/4).ValueOrDie()) {
+      : pool_(SimpleThreadPool::Make(/*threads=*/4).ValueOrDie()) {
     for (int i = 0; i < nfutures; ++i) {
       futures_.push_back(Future<T>::Make());
     }
   }
 
+  // Need to explicilty call the pool shutdown.  The worker threads have a shared_ptr
+  // to the pool so they will keep it alive after this class is destructed.  However, the
```

Comment:
Hmm, does that mean there is a reference cycle which would make the pool immortal? Or just that the pool would be destroyed too late? (hopefully the latter)

Reply:
It is the latter.  The test spawns a task and gives it a reference (via `this` capture) to a future.  The future is destroyed when the test is destroyed which was happening before the spawned task was finished.

Reply:
Actually, it was the former, and that was the ASAN error.  If the worker threads have a strong reference to the thread pool then the thread pool is never destroyed unless it is shutdown (which doesn't work for the eternal pools).  I've fixed it now as worker threads only have a reference to the thread pool's state and not the thread pool (as it was originally).